### PR TITLE
Use `Never` return type to declare function call is guaranteed to end program execution

### PR DESCRIPTION
**Pull request by aidantwoods**
### Summary
Inspired heavily by the introduction of this in [Swift's standard library](https://github.com/apple/swift-evolution/blob/master/proposals/0102-noreturn-bottom-type.md), introduce a non-constructible `Never` type to enforce that functions that are intended to end program execution actually do (not doing so would be a type error).

### Issues Addressed (Optional)
n/a

## Contributor Agreement (Required)

I am submitting this pull request under one or more of the following
licenses:

- [ ] [CC0 - No Rights Reserved](https://creativecommons.org/publicdomain/zero/1.0/)
- [x] [MIT License](https://opensource.org/licenses/MIT)
- [ ] [WTFPL](http://www.wtfpl.net/txt/copying/)

Furthermore, I understand that CMS Airship is released under the GNU Public
License to the general public, as well as private commercial licenses 
(purchasable from [Paragon Initiative Enterprises](https://paragonie.com)).

By submitting this pull request, I acknowledge that my contribution will be
incorporated into CMS Airship, and consent for it to be handled as outlined
above.

(This does not in any way restrict your rights to use your own modifications.
The purpose of this agreement is to maximize awareness and transparency.) 

### Labels Requested

Feel free to list any labels you feel are appropriate for this issue,
and a member of the Airship core team will apply them as soon as 
possible.

* `feedback-sought`
* `documentation`

### CODE REVIEW

File: src/Cabin/Hull/Controller/PublicFiles.php [hunk 1]
```diff
@@ -109,11 +110,13 @@ public function download(string $path, string $default = 'text/plain'): void
             }
             \readfile($realPath);
             exit;
+            return new Never;
```
Comment (aidantwoods) [inline src/Cabin/Hull/Controller/PublicFiles.php]: These statement's are a bit silly, but psalm will complain if they're not there

File: src/Engine/Contract/Never.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
```
Comment (aidantwoods) [inline src/Engine/Contract/Never.php]: This isn't really an interface, so there's probably a better place this could go...

**Comment by paragonie-scott** [comment]
Oh, this is interesting.

**Comment by aidantwoods** [comment]
Incidentally if you know how to make psalm happy without putting return statements in unreachable code let me know (I've tried `@psalm-suppress InvalidReturnType` in the docblock to no avail).

**Pull request merged**